experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,311 @@
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+/// An empty struct to represent "small" type code point tries for the
+///  [`TrieType`] trait. The "fast max" limit is set to 0x0fff.
+pub struct Small;
+
+impl TrieType for Small {
+    const FAST_MAX: u32 = SMALL_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+/// This struct represents a de-serialized CodePointTrie that was exported from
+/// ICU binary data.
+///
+/// For more information:
+/// - [ICU Site design doc](http://site.icu-project.org/design/struct/utrie)
+/// - [ICU User Guide section on Properties lookup](https://unicode-org.github.io/icu/userguide/strings/properties.html#lookup)
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+/// This struct contains the fixed-length header fields of a [`CodePointTrie`].
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    /// Returns a new [`CodePointTrie`] backed by borrowed data for the `index`
+    /// array and `data` array, whose data values have width `W`, for a trie
+    /// type `T`.
+    pub fn try_new(
+        header: CodePointTrieHeader,
+        index: ZeroVec<'trie, u16>,
+        data: ZeroVec<'trie, W>,
+    ) -> Result<CodePointTrie<'trie, W, T>, Error> {
+        if header.data_length < ERROR_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason: "Data array must be large enough to contain error value",
+            });
+        }
+
+        if header.data_length < HIGH_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason:
+                    "Data array must be large enough to contain value for range highStart..U+10FFFF",
+            });
+        }
+
+        if index.len() as u32 != header.index_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of index array does not match corresponding header value",
+            });
+        }
+
+        if data.len() as u32 != header.data_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of data array does not match corresponding header value",
+            });
+        }
+
+        let trie: CodePointTrie<'trie, W, T> = CodePointTrie {
+            header,
+            index,
+            data,
+            _marker_ty: PhantomData,
+        };
+        Ok(trie)
+    }
+
+    /// Returns the position in the data array containing the trie's stored
+    /// error value.
+    fn trie_error_val_index(&self) -> u32 {
```

Comment:
Suggestion: In Rust, I think we should use `usize` instead of `u32` for all index types.  I think the `u32` probably comes from the ICU code.  However, if this is challenging to do, then you can keep it as `u32`.

Reply:
usize:
> The pointer-sized unsigned integer type.
> The size of this primitive is how many bytes it takes to reference any location in memory. For example, on a 32 bit target, this is 4 bytes and on a 64 bit target, this is 8 bytes.

The trie index array entries, and the length of the data array, are at most a little over 18 bits. Definitely don't need more than a u32.